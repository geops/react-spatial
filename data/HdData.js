import Style from 'ol/style/Style';
import Circle from 'ol/style/Circle';
import Stroke from 'ol/style/Stroke';
import Fill from 'ol/style/Fill';
import proj from '../src/Projections';

const { projectionExtent, resolutions } = proj['EPSG:3857'];

export default [
  {
    name: 'Basemap',
    visible: true,
    isBaseLayer: true,
    data: {
      type: 'wmts',
      url:
        `//vtiles.prod.geops.ch/styles/inspirationskarte_de/rendered/` +
        '{TileMatrix}/{TileCol}/{TileRow}.png',
      url2:
        `//vtiles.prod.geops.ch/styles/inspirationskarte_de/rendered/` +
        '{TileMatrix}/{TileCol}/{TileRow}@2x.png',
      url3:
        `//vtiles.prod.geops.ch/styles/inspirationskarte_de/rendered/` +
        '{TileMatrix}/{TileCol}/{TileRow}@3x.png',
      matrixSet: 'webmercator',
      requestEncoding: 'REST',
      projectionExtent,
      resolutions,
    },
    copyright:
      'Â© swisstopo, OpenStreetMap contributors, http://www.imagico.de/, SBB/CFF/FFS 07/2017',
  },
];
