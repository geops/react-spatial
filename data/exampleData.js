import Style from 'ol/style/Style';
import Stroke from 'ol/style/Stroke';
import Fill from 'ol/style/Fill';

export default [
  {
    name: 'OSM Baselayer',
    visible: true,
    isBaseLayer: true,
    data: {
      type: 'xyz',
      url: 'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
    },
  },
  {
    id: 'lineSamples',
    name: 'Lines Samples',
    visible: true,
    data: {
      style: new Style({
        stroke: new Stroke({
          color: '#ffcc33',
          width: 2,
        }),
      }),
      type: 'vectorLayer',
      url:
        'https://raw.githubusercontent.com/openlayers/openlayers/' +
        '3c64018b3754cf605ea19cbbe4c8813304da2539/examples/data/geojson/' +
        'line-samples.geojson',
    },
  },
  {
    id: 'polySamples',
    name: 'Polygons Samples',
    visible: false,
    data: {
      style: new Style({
        stroke: new Stroke({
          color: '#7dff8f',
          width: 3,
        }),
        fill: new Fill({
          color: 'rgba(125, 255, 143, 0.2)',
        }),
      }),
      type: 'vectorLayer',
      url:
        'https://raw.githubusercontent.com/openlayers/openlayers/' +
        '3c64018b3754cf605ea19cbbe4c8813304da2539/examples/data/geojson/' +
        'polygon-samples.geojson',
    },
  },
];
