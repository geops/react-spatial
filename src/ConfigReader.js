import TileLayer from 'ol/layer/Tile';
import XYZ from 'ol/source/XYZ';
import VectorSource from 'ol/source/Vector';
import WMTSSource from 'ol/source/WMTS';
import WMTSTileGrid from 'ol/tilegrid/WMTS';
import GeoJSONFormat from 'ol/format/GeoJSON';
import Layer from './Layer';
import VectorLayer from './VectorLayer';

import projections from './Projections';

class ConfigReader {
  static createXYZLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
      olLayer: new TileLayer({
        zIndex: -1,
        source: new XYZ({
          url: item.data.url,
        }),
      }),
    });
  }

  static createVectorLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new VectorLayer({
      ...conf,
      source: new VectorSource({
        url: item.data.url,
        format: new GeoJSONFormat(),
      }),
      style: item.data.style,
    });
  }

  static createWMTSLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
      olLayer: new TileLayer({
        zIndex: -1,
        source: new WMTSSource({
          url: item.data.url,
          tileGrid: new WMTSTileGrid({
            extent: item.data.projectionExtent,
            resolutions: item.data.resolutions,
            matrixIds: item.data.resolutions.map((res, i) => `${i}`),
          }),
        }),
      }),
    });
  }

  static createEmptyLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
    });
  }

  static createLayer(item) {
    let layer;

    switch (item.data.type) {
      case 'xyz':
        layer = ConfigReader.createXYZLayer(item);
        break;
      case 'wmts':
        layer = ConfigReader.createWMTSLayer(item);
        break;
      case 'vectorLayer':
        layer = ConfigReader.createVectorLayer(item);
        break;
      default:
        layer = ConfigReader.createEmptyLayer(item);
    }

    return layer;
  }

  constructor(map, data, style = null) {
    this.map = map;
    this.data = data;
    this.style = style;

    this.defaults = {
      data: {},
      visible: false,
      isBaseLayer: false,
    };
  }

  loadLayerFromConfig(config) {
    // apply default values
    const item = { ...this.defaults, ...config };

    if (item.data && item.data.type === 'wmts') {
      const proj = projections[item.data.projection];
      if (proj) {
        item.data.projectionExtent = proj.projectionExtent;
        item.data.resolutions = proj.resolutions;
      }
    }

    const layer = ConfigReader.createLayer(item);
    layer.init(this.map);

    if (item.children) {
      item.children.forEach(childConf => {
        layer.addChild(this.loadLayerFromConfig(childConf));
      });
    }

    return layer;
  }

  initialize() {
    return this.data.map(val => this.loadLayerFromConfig(val));
  }
}

const readConfig = (map, data) => {
  const configReader = new ConfigReader(map, data);
  return configReader.initialize();
};

const getVisibleTopic = topicList => {
  let visibleTopic = null;
  topicList.forEach(topic => {
    if (topic.visible) {
      visibleTopic = topic;
    }
  });
  return visibleTopic.children;
};

export default { getVisibleTopic, readConfig };
