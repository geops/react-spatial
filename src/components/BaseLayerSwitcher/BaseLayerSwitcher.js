/* eslint-disable jsx-a11y/interactive-supports-focus */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FaChevronLeft } from 'react-icons/fa';
import { Layer } from 'mobility-toolbox-js/ol';

import './BaseLayerSwitcher.scss';

const propTypes = {
  /**
   * An array of react-spatial layers.
   */
  layers: PropTypes.arrayOf(PropTypes.instanceOf(Layer)).isRequired,

  /**
   * Object containing relative paths to the base layer images. Object
   * keys need to correspond to layer keys
   */
  layerImages: PropTypes.objectOf(PropTypes.string),

  /**
   * CSS class to apply on the container.
   */
  className: PropTypes.string,

  /**
   * Alternative text rendered if layer images can't be loaded
   */
  altText: PropTypes.string,

  /**
   * Button titles.
   */
  titles: PropTypes.shape({
    button: PropTypes.string,
    openSwitcher: PropTypes.string,
    closeSwitcher: PropTypes.string,
  }),

  /**
   * Image (node) rendered in the switcher close button.
   */
  closeButtonImage: PropTypes.node,

  /**
   * Translation function.
   * @param {function} Translation function returning the translated string.
   */
  t: PropTypes.func,
};

const defaultProps = {
  className: 'rs-base-layer-switcher',
  altText: 'Source not found',
  titles: {
    button: 'Base layers',
    openSwitcher: 'Open Baselayer-Switcher',
    closeSwitcher: 'Close Baselayer-Switcher',
  },
  closeButtonImage: <FaChevronLeft />,
  layerImages: undefined,
  t: (s) => s,
};

const getVisibleLayer = (layers) => {
  return layers.find((layer) => layer.visible);
};

const getNextImage = (currentLayer, layers, layerImages) => {
  const currentIndex = layers.indexOf(
    layers.find((layer) => layer === currentLayer),
  );
  const nextIndex = currentIndex + 1 === layers.length ? 0 : currentIndex + 1;
  return layerImages[nextIndex];
};

const getImageStyle = (url) => {
  return url
    ? {
        backgroundImage: `url(${url})`,
        backgroundSize: 'cover',
        backgroundRepeat: 'no-repeat',
        backgroundPosition: 'center',
      }
    : null;
};

function BaseLayerSwitcher({
  layers,
  layerImages,
  className,
  altText,
  titles,
  closeButtonImage,
  t,
}) {
  const baseLayers = layers.filter((layer) => layer.isBaseLayer);
  const [switcherOpen, setSwitcherOpen] = useState(false);
  const [isClosed, setIsClosed] = useState(true);
  const [currentLayer, setCurrentLayer] = useState(
    getVisibleLayer(baseLayers) || baseLayers[0],
  );

  /* Images are loaded from props if provided, fallback from layer */
  const images = layerImages
    ? Object.keys(layerImages).map((layerImage) => layerImages[layerImage])
    : baseLayers.map((layer) => layer.get('previewImage'));

  const openClass = switcherOpen ? ' rs-open' : '';
  const hiddenStyle = switcherOpen && !isClosed ? 'visible' : 'hidden';

  const handleSwitcherClick = () => {
    if (baseLayers.length === 2) {
      /* On only two layer options the opener becomes a layer toggle button */
      const nextLayer = baseLayers.find((layer) => !layer.visible);
      currentLayer.setVisible(false);
      setCurrentLayer(nextLayer);
      nextLayer.setVisible(true);
      return;
    }
    // eslint-disable-next-line consistent-return
    return setSwitcherOpen(true) && setIsClosed(false);
  };

  const onLayerSelect = (layer) => {
    if (!switcherOpen) {
      setSwitcherOpen(true);
      return;
    }
    setCurrentLayer(layer);
    layer.setVisible(true);
    baseLayers.filter((l) => l !== layer).forEach((l) => l.setVisible(false));
    setSwitcherOpen(false);
  };

  /* Get next image for closed button */
  const nextImage = getNextImage(currentLayer, baseLayers, images);

  useEffect(() => {
    /* Ensure correct layer is active on app load */
    if (currentLayer !== getVisibleLayer(baseLayers)) {
      setCurrentLayer(getVisibleLayer(baseLayers) || baseLayers[0]);
    }
  }, [currentLayer, baseLayers]);

  useEffect(() => {
    /* Used for correct layer image render with animation */
    let timeout;
    if (!switcherOpen) {
      timeout = setTimeout(() => {
        setIsClosed(true);
      }, 200);
    } else {
      timeout = setTimeout(() => {
        setIsClosed(false);
      }, 800);
    }
    return () => clearTimeout(timeout);
  }, [switcherOpen]);

  if (!baseLayers || baseLayers.length < 2 || !currentLayer) {
    return null;
  }

  const toggleBtn = (
    <div className="rs-base-layer-switcher-btn-wrapper">
      <div
        className="rs-base-layer-switcher-close-btn"
        role="button"
        onClick={() => setSwitcherOpen(false)}
        onKeyPress={(e) => e.which === 13 && setSwitcherOpen(false)}
        tabIndex={switcherOpen ? '0' : '-1'}
        aria-label={titles.closeSwitcher}
        title={titles.closeSwitcher}
      >
        {closeButtonImage}
      </div>
    </div>
  );

  return (
    <div className={`${className}${openClass}`}>
      {
        <div
          className={`rs-base-layer-switcher-button rs-opener${openClass}`}
          role="button"
          title={titles.openSwitcher}
          aria-label={titles.openSwitcher}
          onClick={handleSwitcherClick}
          onKeyPress={(e) => {
            if (e.which === 13) {
              handleSwitcherClick();
            }
          }}
          style={getImageStyle(nextImage)}
          tabIndex="0"
        >
          <div className="rs-base-layer-switcher-title">
            {baseLayers.length !== 2
              ? titles.button
              : baseLayers.find((layer) => !layer.visible) &&
                t(baseLayers.find((layer) => !layer.visible).name)}
          </div>
          {nextImage ? null : <span className="rs-alt-text">{t(altText)}</span>}
        </div>
      }
      {baseLayers.map((layer, idx) => {
        const layerName = layer.name;
        const activeClass = layerName === currentLayer.name ? ' rs-active' : '';
        const imageStyle = getImageStyle(
          layerImages ? layerImages[`${layer.key}`] : layer.get('previewImage'),
        );
        return (
          <div
            key={layer.key}
            className="rs-base-layer-switcher-btn-wrapper"
            style={{
              overflow: hiddenStyle,
              zIndex: baseLayers.length - idx,
            }}
          >
            <div
              className={`rs-base-layer-switcher-button${openClass}`}
              role="button"
              title={t(layerName)}
              aria-label={t(layerName)}
              onClick={() => onLayerSelect(layer)}
              onKeyPress={(e) => {
                if (e.which === 13) {
                  onLayerSelect(layer);
                }
              }}
              style={imageStyle}
              tabIndex={switcherOpen ? '0' : '-1'}
            >
              <div className={`rs-base-layer-switcher-title${activeClass}`}>
                {t(layerName)}
              </div>
              {imageStyle ? null : (
                <span className="rs-alt-text">{t(altText)}</span>
              )}
            </div>
          </div>
        );
      })}
      {toggleBtn}
    </div>
  );
}

BaseLayerSwitcher.propTypes = propTypes;
BaseLayerSwitcher.defaultProps = defaultProps;

export default BaseLayerSwitcher;
