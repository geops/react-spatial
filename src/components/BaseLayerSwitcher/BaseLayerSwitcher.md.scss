.rs-base-layer-example {
  position: relative;

  .rs-base-layer-switcher {
    position: absolute;
    bottom: 10px;
    left: 10px;
  }
}
