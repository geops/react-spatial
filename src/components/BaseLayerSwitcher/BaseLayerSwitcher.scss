.rs-base-layer-switcher {
  position: relative;
  width: 100px;
  transition: 800ms width;
  overflow: hidden;
  display: flex;
  padding: 2px 2px;
  pointer-events: none;

  &.rs-open {
    width: 600px;
  }

  .rs-base-layer-switcher-button {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-end;
    height: 80px;
    width: 100px;
    min-width: 100px;
    margin-right: 4px;
    background-color: rgb(197, 197, 197);
    border: 2px solid white;
    box-sizing: border-box;
    pointer-events: auto;

    &.rs-opener {
      position: absolute;
      top: 2px;       // For IE
      left: 2px;        // For IE
      opacity: 1;
      z-index: 100;
      transition: 700ms opacity, 1000ms z-index;

      &.rs-open {
        opacity: 0;
        z-index: 0;
      }
    }
  }

  .rs-alt-text {
    flex-grow: 1;
    display: flex;
    justify-content: center;
    text-align: center;
    align-items: center;
    max-width: 80px;
    overflow: hidden;
  }

  .rs-base-layer-switcher-btn-wrapper {
    display: flex;
    align-items: center;
  }

  .rs-base-layer-switcher-close-btn {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background-color: white;
    width: 24px;
    height: 24px;
    border-radius: 50px;
    pointer-events: auto;
  }

  .rs-open {
    &:focus,
    &:hover {
      outline: none;
      border: 2px solid red;
    }
  }

  .rs-base-layer-switcher-title {
    width: 100%;
    max-height: 12px;
    font-size: 10px;
    padding: 2px 0;
    color: white;
    background-color: rgba(0, 0, 0, 0.7);
    text-align: center;

    &.rs-active {
      font-weight: bold;
    }
  }
}
