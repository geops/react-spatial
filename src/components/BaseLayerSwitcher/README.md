#

This demonstrates the use of BaseLayerSwitcher.

```jsx
import React from 'react';
import Map from 'ol/Map';
import BaseLayerSwitcher from 'react-spatial/components/BaseLayerSwitcher';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import osmImage from 'react-spatial/images/baselayer/osm.baselayer.png';
import osmhotImage from 'react-spatial/images/baselayer/osm.baselayer.hot.png';
import openTopoImage from 'react-spatial/images/baselayer/open.topo.map.png';

const center = [1149722.7037660484, 6618091.313553318];
const map = new Map({ controls: [] });
const layers = ConfigReader.readConfig(treeData);
const layerImages = {
  'osm.baselayer': osmImage,
  'osm.baselayer.hot': osmhotImage,
  'open.topo.map': openTopoImage,
};

<div className="rs-base-layer-example">
  <BasicMap
    map={map}
    center={center}
    zoom={6}
    layers={layers}
    tabIndex={0}
  />
  <BaseLayerSwitcher
    layers={layers}
    layerImages={layerImages}
  />
</div>;
```
