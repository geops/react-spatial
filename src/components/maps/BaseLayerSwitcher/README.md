#

This demonstrates the use of BaseLayerSwitcher.

```jsx
import React from 'react';
import OLMap from 'ol/Map';
import BaseLayerSwitcher from 'react-spatial/components/maps/BaseLayerSwitcher';
import BasicMap from 'react-spatial/components/maps/BasicMap';
import LayerService from 'react-spatial/LayerService';
import ConfigReader from 'react-spatial/ConfigReader';
import osmImage from '../../../images/baselayer/osm.baselayer.png';
import osmhotImage from '../../../images/baselayer/osm.baselayer.hot.png';
import openTopoImage from '../../../images/baselayer/open.topo.map.png';

const center = [1149722.7037660484, 6618091.313553318];
const map = new OLMap({ controls: [] });
const layers = ConfigReader.readConfig(treeData);
const layerService = new LayerService(layers);
const layerImages = {
  'osm.baselayer': osmImage,
  'osm.baselayer.hot': osmhotImage,
  'open.topo.map': openTopoImage,
};

<div className="rs-base-layer-example">
  <BasicMap
    map={map}
    center={center}
    zoom={6}
    layers={layers}
    tabIndex={0}
  />
  <BaseLayerSwitcher
    layers={layers}
    layerImages={layerImages}
  />
</div>;
```
