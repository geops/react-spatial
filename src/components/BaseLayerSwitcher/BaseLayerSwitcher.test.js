/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';
import renderer from 'react-test-renderer';
import ConfigReader from '../../ConfigReader';
import data from '../../../data/TreeData';
import BaseLayerSwitcher from './BaseLayerSwitcher';

configure({ adapter: new Adapter() });

const getSnapshot = newData => {
  const layers = ConfigReader.readConfig(newData || data);
  const layerImages = {
    layer1: 'foo',
  };
  const component = renderer.create(
    <BaseLayerSwitcher layers={layers} layerImages={layerImages} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

const mountComp = newData => {
  const layers = ConfigReader.readConfig(newData || data);
  const layerImages = {
    layerFoo: 'foo',
    layerBar: 'bar',
    layerFoobar: 'foobar',
  };
  return mount(<BaseLayerSwitcher layers={layers} layerImages={layerImages} />);
};

describe('BaseLayerToggler', () => {
  describe('matches snapshots', () => {
    test('using default properties.', () => {
      getSnapshot();
    });
  });

  test('the correct baselayer is visible on mount', () => {
    const comp = mountComp(data);
    expect(comp.props().layers[0].getVisible()).toBe(true);
  });

  test('adds close button when opening the switcher', () => {
    const comp = mountComp(data);
    expect(comp.find('.rs-base-layer-switcher-button').length).toBe(1);
    comp
      .find('.rs-base-layer-switcher-button')
      .at(0)
      .simulate('click');
    expect(comp.find('.rs-base-layer-switcher-close-btn').length).toBe(1);
  });

  test('removes open class and switches layer on click', () => {
    const comp = mountComp(data);
    comp
      .find('.rs-base-layer-switcher-button')
      .at(0)
      .simulate('click');
    comp
      .find('.rs-base-layer-switcher-button')
      .at(2)
      .simulate('click');
    expect(
      comp
        .props()
        .layers.filter(layer => layer.getIsBaseLayer())[2]
        .getVisible(),
    ).toBe(true);
    expect(comp.find('.rs-base-layer-switcher rs-open').exists()).toBe(false);
  });
});
