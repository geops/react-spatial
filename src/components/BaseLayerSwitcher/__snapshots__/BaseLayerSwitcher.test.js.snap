// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLayerToggler matches snapshots using default properties. 1`] = `
<div
  className="rs-base-layer-switcher"
>
  <div
    aria-label="Source not found"
    className="rs-base-layer-switcher-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "backgroundImage": "url(undefined)",
        "backgroundPosition": "center",
        "backgroundRepeat": "no-repeat",
        "backgroundSize": "cover",
      }
    }
    tabIndex="0"
    title="Open Baselayer-Switcher"
  >
    <div
      className="rs-base-layer-switcher-title rs-closed"
    >
      Base layers
    </div>
    <span
      className="rs-alt-text"
    >
      Source not found
    </span>
  </div>
</div>
`;
